File: repeateddna.java
// Two HashSet Solution
// Time complexity: O(N)
// Space complexity: O(N)
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        if (s == null || s.length() <= 10) return new ArrayList<>();
        final int N = s.length();
        Set<String> repeatSet = new HashSet<>();
        Set<String> seen = new HashSet<>();
        for (int i = 0; i <= N - 10; i++) {
            String substr = s.substring(i, i + 10);
            if (!seen.add(substr)) {
                repeatSet.add(substr);
            }
        }
        return new ArrayList<>(repeatSet);
    }
}

File: repeatedDNA.py
#Time: O(N)
#Space: O(N)

class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        if len(s)  == 0:
            return []
        rep = set()
        result = set()
        
        for i in range(len(s)-9):
            t  = s[i:i+10]
            if t in rep:
                result.add(t)
            rep.add(t)
            
        return result